lib/srv/alpnproxy/proxy.go

```diff
@@ -359,43 +361,64 @@ func (p *Proxy) handleConn(ctx context.Context, clientConn net.Conn) error {
 		return trace.Wrap(err)
 	}
 
+	// Wrap incoming net.Conn with waitConn in case we have to wait for aysnc
+	// handlers.
+	waitConn := newWaitConn(ctx, conn)
+	conn = waitConn
+
+	// Upgrade conn to tls.Conn if not forwarding TLS.
+	if !handlerDesc.ForwardTLS {
+		conn, handlerDesc, err = p.setupTLSConnForHandler(conn, handlerDesc, opts)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	// Close clientConn for "sync" handlers. "Async" handlers will close the
+	// connections by themselves.
+	if !handlerDesc.IsAsync {
+		defer func() {
+			if err := clientConn.Close(); err != nil && !utils.IsOKNetworkError(err) {
+				p.log.WithError(err).Warnf("Failed to close client connection.")
+			}
+		}()
+	}
+	// Wait for async handlers if options says so.
+	if handlerDesc.IsAsync && opts.waitForAsyncHandlers {
+		defer waitConn.WaitForClose()
+	}
```

Comment:
Wonder if it it possible to to combine `ListenerMuxWrapper` and waitConn so `ListenerMuxWrapper) HandleConnection` can cast incoming connection  waitConn and wait till `waitConn.WaitForClose()`